mnemonics/src/main/java/org/kethereum/bip39/Mnemonic.kt

```diff
@@ -0,0 +1,175 @@
+     * The mnemonic phrase is given as a list of words and the seed can be salted using a password
+     */
+    fun mnemonicToSeed(phrase: String, password: String = ""): ByteArray {
+        return mnemonicToSeed(phrase.split(" ").toTypedArray(), password)
+    }
+
+    /**
+     * Generates a seed buffer from a mnemonic phrase according to the BIP39 spec.
+     * The mnemonic phrase is given as a list of words and the seed can be salted using a password
+     */
+    fun mnemonicToSeed(words: Array<String>, password: String = ""): ByteArray {
+        val pass = words.joinToString(" ")
+        val salt = "mnemonic" + password
+
+        val keyFactory = SecretKeyFactory.getInstance("PBKDF2withHmacSHA512")
+        val spec = PBEKeySpec(pass.toCharArray(), salt.toByteArray(), 2048, 512)
+        val seed = keyFactory.generateSecret(spec).encoded
+        return seed
+    }
+
+    /**
+     * Converts a phrase (list of words) into a [ByteArray] entropy buffer according to the BIP39 spec
+     */
+    fun mnemonicToEntropy(phrase: String): ByteArray {
+        return mnemonicToEntropy(phrase.split(" ").toTypedArray())
+    }
+
+    /**
+     * Converts a list of words into a [ByteArray] entropy buffer according to the BIP39 spec
+     */
+    fun mnemonicToEntropy(words: Array<String>): ByteArray {
+        if (words.size % 3 > 0)
+            throw RuntimeException("Word list size must be multiple of three words.")
+
+        if (words.isEmpty())
+            throw RuntimeException("Word list is empty.")
+
+        val numTotalBits = words.size * 11
+        val bitArray = BooleanArray(numTotalBits)
+
+        for ((phraseIndex, word) in words.withIndex()) {
+
+            val dictIndex = Collections.binarySearch(ENGLISH, word)
+            if (dictIndex < 0)
+                throw RuntimeException("word($word) not in known word list")
+
+            // Set the next 11 bits to the value of the index.
+            for (bit in 0..10)
+                bitArray[phraseIndex * 11 + bit] = dictIndex and (1 shl (10 - bit)) != 0
+        }
+
+        val numChecksumBits = numTotalBits / 33
+        val numEntropyBits = numTotalBits - numChecksumBits
+
+        val entropy = bitsToBytes(bitArray, numEntropyBits / 8 )
+
+        // Take the digest of the entropy.
+        val hash = entropy.sha256()
+        val hashBits = bytesToBits(hash)
+
+        // Check all the checksum bits.
+        for (i in 0 until numChecksumBits)
+            if (bitArray[numEntropyBits + i] != hashBits[i])
+                throw RuntimeException("mnemonic checksum does not match")
+
+        return entropy
+    }
+
+    fun mnemonicToKey(phrase : String, path : String, saltPhrase: String = "") : ExtendedKey {
+        val generatedSeed = Mnemonic.mnemonicToSeed(phrase, saltPhrase)
+        return BIP32.generateKey(generatedSeed, path)
+    }
+
+    private fun bytesToBits(data: ByteArray): BooleanArray {
```

Comment:
think bytesToBits and bitsToBytes should not live here but in e.g. the extensions module

Reply:
I moved them and converted to extension functions.